### DISCUSSION TITLE
'secret-encryption' feature in k3s

### DISCUSSION THREAD

**Discussion by mallikarjunraju** [k3s-io/k3s#3870] [Q&A]
Hi All,

I am trying to implement 'secret-encryption' feature in k3s.

With below command, I can enable it at k3s installation time.

`curl -sfL https://get.k3s.io | INSTALL_K3S_VERSION="v1.21.1+k3s1" K3S_KUBECONFIG_MODE="644" INSTALL_K3S_EXEC="--secrets-encryption" sh -`

Also found that, the secret key (aescbc encrypted) is located at  "`/var/lib/rancher/k3s/server/cred/encryption-config.json`"

My aim is to store aescbc key safely. (In KMS store or TPM chip)

**Question 1:** I tried to update the k3s-server api after k3s installation WITHOUT 'secret-encryption' by using below command 
   
`k3s server --kube-apiserver-arg='encryption-provider-config=/etc/encryption-config.yaml'  `

where I was able to use my custom config yaml. But this command is **not exiting** and encrypting 

'`/var/lib/rancher/k3s/server/db/state.db-wal`' rather than '`/var/lib/rancher/k3s/server/db/state.db`'

How Can I use my custom config yaml with my own aescbc key and why that command is not exiting and encrypting different db?

**Question 2:** Anyone tried KMS in k3s for secret-encryption?

Any documentation links can be helpful. Below link not that helpful.
https://rancher.com/docs/k3s/latest/en/installation/install-options/server-config/

**Comment by mallikarjunraju** [comment]
As a workaround, I was able to apply the secret encryption at the time of k3s installation itself.

`curl -sfL https://get.k3s.io | sh -s - --kube-apiserver-arg='encryption-provider-config=/home/raju/secretConfig.yaml' `


if anyone apply secret encryption after plain k3s installation by updating k3s server. Please share.

**Reply by VivekSheshadari** [reply]
facing same issue, basically i am using k3d with k3s image and not able to achive encryption at rest

**Reply by brandond** [reply]
@VivekSheshadari it seems you have some misconceptions about what "secrets encryption" does. https://github.com/k3s-io/k3s/issues/9559#issuecomment-1961711137

**Comment by dereknola** [comment]
Two things happening here:
- `--secrets-encryption` server flag: This enable the [K3s management of secrets-encryption ](https://docs.k3s.io/cli/secrets-encrypt#secrets-encryption-tool)and enables secrets encryption config on the kube-apiserver. Once you have started a K3s server with this, the only way to turn it off is to use the [k3s secrets-encrypt disable](https://docs.k3s.io/cli/secrets-encrypt#secrets-encryption-disablere-enable) command.

- `--kube-apiserver-arg='encryption-provider-config`: this is how you enable just [K8s secrets encryption at rest](https://kubernetes.io/docs/tasks/administer-cluster/encrypt-data/). This would also be how you use KMS for K3s. Note that if you go down this road, K3s has no control or insight into your secrets-encryption, and any `k3s secrets-encrypt` commands are not guaranteed to work.
